### DISCUSSION TITLE
[Other] Upgrade path question

### DISCUSSION THREAD

**Discussion by TimKSwe** [rabbitmq/rabbitmq-server#12767] [Other] [answered]
### Community Support Policy

- [X] I have read [RabbitMQ's Community Support Policy](https://github.com/rabbitmq/rabbitmq-server/blob/main/COMMUNITY_SUPPORT.md)

### RabbitMQ version used

3.13.7 or older

### How is RabbitMQ deployed?

Other

### Steps to reproduce the behavior in question

This isn't a behavior question.

We're currently running 3.11.9 (!) and obviously need to upgrade. I have a question about the upgrade path.

The table on this page: https://www.rabbitmq.com/docs/upgrade#rabbitmq-version-upgradability suggests that, in order to upgrade to 3.12.x, you have to be on 3.11.18. Am I right in assuming that we would need to upgrade as follows:

3.11.9 --> 3.11.18 --> 3.12.x --> 3.13.x. --> 4.0.x

...or am I misinterpreting the table?

In this question, I am consciously ignoring the potential existence of breaking changes. I'll need to investigate those separately. Any information that anyone would like to volunteer about such changes would of course be welcome! :-)

**Comment by mkuratczyk** [comment, marked as answer]
Correct. And don't forget to enable all stable feature flags before/after attempting an upgrade (if they are already enabled, it's a no-op, so safe to run the command both before and after).

Do NOT enable khepri_db feature flag on 3.13. It would make it non-upgradeable to 4.0 (it was experimental only).
On 4.0, you can enable khepri_db - it's optional, but support starting in 4.0.

**Reply by michaelklishin** [reply]
Alternatively, instead of a series of upgrades, you can [use a Blue/Green upgrade](https://www.rabbitmq.com/docs/blue-green-upgrade) and move directly to `4.0.x` that way. It is more involved but works even for multi-minor gaps, e.g. 3.8.x => 4.0.x upgrades.

**Comment by TimKSwe** [comment]
Thank you! I had also seen the tips surrounding Khepri support, but it's good to have that reminder. :-) (I also don't know what Khepri is -- heck, I'm only just now learning the details of RabbitMQ because I volunteered to take it over for my site....)

**Reply by mkuratczyk** [reply]
It's a replacement for a critical internal component (a database for storing metadata about queues and other things):
https://www.rabbitmq.com/docs/metadata-store/how-to-enable-khepri

It will become the default in the future and will become the only option at some point. 4.0 is the first version where it's considered stable.
